Sorted2

lundman opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

lundman committed:
macOS: Add macOS support

lundman committed:
Upstream: Add macOS to autoconf

lundman committed:
Upstream: Do not -lrt library on macOS

lundman committed:
Upstream: Add macOS mount commands for setuid

lundman committed:
Upstream: DKICEJECT is defined in macOS

lundman committed:
Upstream: add BUILD_MACOS to libzutil

lundman committed:
Upstream: Include sysmacros.h for ISP2

lundman committed:
Upstream: more Makefile.am changes

lundman committed:
Upstream: define zfs_file_t for macOS

Signed-off-by: Jorgen Lundman <lundman@lundman.net>

lundman committed:
Upstream: add ASSERTV

Not actually used by upstream, the theory here being they are
not offended of it existing, even if they do not want to use it.

lundman committed:
Upstream: Silence used-uninitialized warning

In the default: case.

lundman committed:
Upstream: Add APPLE zpl_attr_t entries

Tempted to use #ifdef __APPLE__ but it appears to not hurt other
platforms.

lundman committed:
Upstream: nvpair_alloc_spl uses vmem_alloc()

Referring to the Linux vmem_alloc(), illumos already has one
which we handle in zfs_context_os.h

lundman committed:
Upstream: move icp EMPTY_TASKQ into taskq.h

Guessing a little on the Linux side, make sure this still compiles there.

lundman committed:
Upstream: Add macOS zfs_prop

lundman committed:
Upstream: macOS ioctls

lundman committed:
Upstream: Add macOS member to spa_t

lundman committed:
Upstream: Add macOS section to zfs_xattr_owner_unlinked

lundman committed:
Upstream: expose zfs_ioctl_register_pool()

It appears FreeBSD did the same for zfs_ioctl_register_dataset_nolog()
as they use it, so following suit for zfs_ioctl_register_pool()

lundman committed:
Upstream: port macOS libefi

lundman committed:
Upstream: Makefiles add libefi

lundman committed:
Upstream: additional changes to macos properties

lundman committed:
Upstream: additional changes to macOS SA changes

lundman committed:
Upstream: macOS can also set va_type

Honestly, don't know if this is required, but XNU does like
va_type and va_mode

lundman committed:
Upstream: additional mntent.h entries

lundman committed:
Upstream: macOS default mount is /Volumes

lundman committed:
Upstream: allow spa_create_os()  squash

macOS needs to maintain an IOkit proxy node for pools, and know when
to init/fini. This could be made a little more generic if that is
preferred.

lundman committed:
Upstream: add IO calls for iokit

Is this the best way? We could add ", func, private" to the
existing IO, and either send by uio, or by func(private).

lundman committed:
Upstream: Makefile squash

lundman committed:
Upstream: Add APPLE to headers squash

lundman committed:
Upstream: zfs-tests to support macOS

Additional work, test-runner.py assumes it can dlopen("librt.so.1")
because it needs clock_gettime(). We create a librt.so.1 in tests/
and DYLD_LIBRARY_PATH to it.

lundman committed:
Upstream: Allow cmd/zfs mount unmount of snapshots

lundman committed:
Upstream: Add snapshot events

lundman committed:
Upstream: squash with Makefiles

lundman committed:
Upstream: Warning about DEBUG build of dsl_scan.c lua/lvm.c

This is an APPLE issue, it easily overflows the stack when built with
-O0, as the frame grows huge.

A better solution needs to be found, are there other tweaks that
can be done so the frame doesn't balloon.

lundman committed:
Upstream: add macOS to abd_impl.h

lundman committed:
Upstream: libzfs/Makefile change (squash)

lundman committed:
Upstream: pass handle to do_unmount()

Apparently upstream has already made this change for mount, so only
unmount remains.

OSX wants it for devdisks, and handling of snapshots.
How unattractive is it with the type difference between mount, and
unmount?

lundman committed:
Upstream: abd keep orig_size for free

lundman committed:
macOS: change update_pages

OsX traditionally used uio_copy to re-issue the write in update_pages() to
mapped memory. Based on Linux, it is now changed to using dmu_read() from
ARC, to save the uio allocation, and clean up zfs_write().

Commit:
zfsvfs_setup(): zap_stats_t may have undefined content when accessed

Signed-off-by: Allan Jude <allanjude@klarasystems.com>

lundman committed:
Upstream: AC_DEFINE KERNEL_MODPREFIX configure [squash]

lundman committed:
macOS: enhance getxattr/setxattr code to use txg

lundman committed:
Upstream: move asm_linkage.h to platform folder

Move the:

module/icp/include/sys/ia32/asm_linkage.h
  ->
module/os/$platform/spl/include/ia32/sys/asm_linkage.h
lib/libspl/include/os/linux/ia32/sys/asm_linkage.h

Moving the asm_linkage.h to platform headers allow for easier
modules/icp compile, as well as, the additional assembler work,
including setjmp.S

Move assembler source files into os/linux/ as they are not compatible
with macOS clang. Appears to be a bracket usage difference, it is unknown
if gcc would be happy with clang's style.

Future PR will bring in os/macos/ files.

Give FreeBSD an asm_linkage.h header to include asmacros.h, and
simplify atomic.S. It is unknown if the kernel part if used.

Update the top header:
    module/icp/include/sys/asm_linkage.h
to include "ia32/sys/asm_linkage.h" instead of "sys/ia32/". As far
as I can tell, there never was a "sys/ia32", always "ia32/sys".

This is mostly moving source files around and updating Makefiles, which
naturally means there is a high chance for mistakes.

Signed-off-by: Jorgen Lundman <lundman@lundman.net>

lundman committed:
Upstream: macOS files to asm_linkage.h PR [Makefile squash]

lundman committed:
macOS: xattr writes to use txg, and correct xattr=sa

lundman committed:
Upstream: configure changes [squash]

lundman committed:
Upstream: cmd/zfs rollback [squash]

lundman committed:
Upstream: lib/libzfs/Makefile changes [squash]

lundman committed:
macOS: rollback to notify finder of change

lundman committed:
Upstream: Makefile [squash]

lundman committed:
macOS: setattr(uid/gid) incorrect

lundman committed:
Upstream: reverts recent configure.ac change [squash]

lundman committed:
Upstream: reverts recent Makefile change [squash]

lundman committed:
macOS: replace obj-C function with C

lundman commented on lib/libzfs/os/macos/libzfs_util_os.c:
```diff
@@ -333,7 +333,37 @@ execvpe(const char *name, char * const argv[], char * const envp[])
+	id workspace = ((id(*)(Class,SEL))objc_msgSend)(NSWorkspace,
+	    sharedWorkspace);
+	((id(*)(id,SEL,id))objc_msgSend)(workspace, noteFileSystemChanged, ns_path);
+}
```
@burito I shamelessly stole this section from your sources, let me know if you want any credits? :)

burito replied:
Thanks for tagging me, so I know that someone found my code useful :-)
If you want to credit me that's nice, but I wouldn't insist on it. Looks to me like I just served as a syntax guide. Which I should pass on, while I managed to figure out basic function calling by myself, the code that really cleared everything up for me, and which I heavily based mine on, was https://github.com/jimon/osx_app_in_plain_c - worth checking out if you want to do something more complicated.

lundman committed:
macOS: move AT_ to ATTR_

Appears Linux already defines them in headers, but doesn't use them.
Since we just match to XNU values, it doesn't matter what they are called.

lundman committed:
macOS: zfs_writes were truncated

alaviss committed:
mkfile: include missing headers

Without these headers, compilation fails on musl libc with offset_t
being undeclared and MIN being implictly declared.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Hiếu Lê <leorize+oss@disroot.org>
Closes #10406

behlendorf committed:
ztest: Fix spa_open() ENOENT failures

The pool may not be imported when the previous pass is terminated.
In which case, spa_open() will return ENOENT to indicate the pool
is not currently imported.  Refactor to code slightly to handle
this case by importing the pool and then retrying the spa_open().

The ztest_import() function was moved before ztest_run() and the
import logic split in to a small internal helper function.  The
ztest_freeze() function was also moved but no changes were made.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
Closes #10407

Commit:
Improve compatibility with C++ consumers

C++ is a little picky about not using keywords for names, or string
constness.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Jorgen Lundman <lundman@lundman.net>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #10409

lundman committed:
macOS: const required after C++ fix commit

lundman committed:
Replace sprintf()->snprintf() and strcpy()->strlcpy()

The strcpy() and sprintf() functions are deprecated on some platforms.
Care is needed to ensure correct size is used.  If some platforms
miss snprintf, we can add a #define to sprintf, likewise strlcpy().

The biggest change is adding a size parameter to zfs_id_to_fuidstr().

The various *_impl_get() functions are only used on linux and have
not yet been updated.

Reviewed by: Sean Eric Fagan <sef@ixsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10400

nivedita76 committed:
Cleanup linux module kbuild files

The linux module can be built either as an external module, or compiled
into the kernel, using copy-builtin. The source and build directories
are slightly different between the two cases, and currently, compiling
into the kernel still refers to some files from the configured ZFS
source tree, instead of the copies inside the kernel source tree. There
is also duplication between copy-builtin, which creates a Kbuild file to
build ZFS inside the kernel tree, and the top-level module/Makefile.in.

Fix this by moving the list of modules and the CFLAGS settings into a
new module/Kbuild.in, which will be used by the kernel kbuild
infrastructure, and using KBUILD_EXTMOD to distinguish the two cases
within the Makefiles, in order to choose appropriate include
directories etc.

Module CFLAGS setting is simplified by using subdir-ccflags-y (available
since 2.6.30) to set them in the top-level Kbuild instead of each
individual module. The disabling of -Wunused-but-set-variable is removed
from the lua and zfs modules. The variable that the Makefile uses is
actually not defined, so this has no effect; and the warning has long
been disabled by the kernel Makefile itself.

The target_cpu definition in module/{zfs,zcommon} is removed as it was
replaced by use of CONFIG_SPARC64 in
  commit 70835c5b755e ("Unify target_cpu handling")

os/linux/{spl,zfs} are removed from obj-m, as they are not modules in
themselves, but are included by the Makefile in the spl and zfs module
directories. The vestigial Makefiles in os and os/linux are removed.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10379
Closes #10421

nivedita76 committed:
Fix VPATH builds for user config

cmd/zpool and lib/libzutil Makefile's use -I., which won't work with a
VPATH build. Replace it with -I$(srcdir) instead.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10379
Closes #10421

lundman committed:
Upstream: sys/fm/fs/zfs.h change [squash]

lundman committed:
macOS: Add primitive auto unmount of idle snapshots.

Commit:
Don't erase final byte of envblock

When we copy the envblock's contents out, we currently treat it as 
a normal C string. However, this functionality is supposed to more
closely emulate interacting with a file. As a consequence, we were 
incorrectly truncating the contents of the envblock by replacing 
the final byte of the buffer with a null character.

Reviewed-by: Pavel Zakharov <pavel.zakharov@delphix.com>
Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Paul Dagnelie <pcd@delphix.com>
Closes #10405

pjd committed:
Remove redundant includes

By removing excessive includes it takes us a small step close to
compiling this file in userland.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by:	Pawel Jakub Dawidek <pawel@dawidek.net>
Closes #10415

pjd committed:
Restore support for in-kernel ZFS ioctls

In Illumos it is possible to call ioctl functions from within the
kernel by passing the FKIOCTL flag. Neither FreeBSD nor Linux support
that, but it doesn't hurt to keep it around, as all the code is there.

Before this commit it was a dead code and zc_iflags was always zero.
Restore this functionality by allowing to pass a flag to the
zfsdev_ioctl_common() function.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by:	Pawel Jakub Dawidek <pawel@dawidek.net>
Closes #10417

c0d3z3r0 committed:
Linux 5.8 compat: __vmalloc()

The `pgprot` argument has been removed from `__vmalloc` in Linux 5.8,
being `PAGE_KERNEL` always now [1].

Detect this during configure and define a wrapper for older kernels.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/mm/vmalloc.c?h=next-20200605&id=88dca4ca5a93d2c09e5bbc6a62fbfc3af83c4fca

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Co-authored-by: Sebastian Gottschall <s.gottschall@dd-wrt.com>
Co-authored-by: Michael Niewöhner <foss@mniewoehner.de>
Signed-off-by: Sebastian Gottschall <s.gottschall@dd-wrt.com>
Signed-off-by: Michael Niewöhner <foss@mniewoehner.de>
Closes #10422

c0d3z3r0 committed:
Move GFP flags kernel compatibility code

Move the GFP flags kernel compat code from c file to kmem header.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Michael Niewöhner <foss@mniewoehner.de>
Closes #10424

lundman committed:
macOS: uio changes after uio PR

lundman committed:
Upstream: Makefile change [squash]

gamanakis committed:
Trim L2ARC

The l2arc_evict() function is responsible for evicting buffers which
reference the next bytes of the L2ARC device to be overwritten. Teach
this function to additionally TRIM that vdev space before it is
overwritten if the device has been filled with data. This is done by
vdev_trim_simple() which trims by issuing a new type of TRIM,
TRIM_TYPE_SIMPLE.

We also implement a "Trim Ahead" feature. It is a zfs module parameter,
expressed in % of the current write size. This trims ahead of the
current write size. A minimum of 64MB will be trimmed. The default is 0
which disables TRIM on L2ARC as it can put significant stress to
underlying storage devices. To enable TRIM on L2ARC we set
l2arc_trim_ahead > 0.

We also implement TRIM of the whole cache device upon addition to a
pool, pool creation or when the header of the device is invalid upon
importing a pool or onlining a cache device. This is dependent on
l2arc_trim_ahead > 0. TRIM of the whole device is done with
TRIM_TYPE_MANUAL so that its status can be monitored by zpool status -t.
We save the TRIM state for the whole device and the time of completion
on-disk in the header, and restore these upon L2ARC rebuild so that
zpool status -t can correctly report them. Whole device TRIM is done
asynchronously so that the user can export of the pool or remove the
cache device while it is trimming (ie if it is too slow).

We do not TRIM the whole device if persistent L2ARC has been disabled by
l2arc_rebuild_enabled = 0 because we may not want to lose all cached
buffers (eg we may want to import the pool with
l2arc_rebuild_enabled = 0 only once because of memory pressure). If
persistent L2ARC has been disabled by setting the module parameter
l2arc_rebuild_blocks_min_l2size to a value greater than the size of the
cache device then the whole device is trimmed upon creation or import of
a pool if l2arc_trim_ahead > 0.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Adam D. Moss <c@yotes.com>
Signed-off-by: George Amanakis <gamanakis@gmail.com>
Closes #9713
Closes #9789 
Closes #10224

ikozhukhov committed:
ZTS: Fix add-o_ashift.ksh

Use option '-o' after action for compatibility

Reviewed-by: George Melikov <mail@gmelikov.ru>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Signed-off-by: Igor Kozhukhov <igor@dilos.org>
Closes #10426

ahrens committed:
File incorrectly zeroed when receiving incremental stream that toggles -L

Background:

By increasing the recordsize property above the default of 128KB, a
filesystem may have "large" blocks.  By default, a send stream of such a
filesystem does not contain large WRITE records, instead it decreases
objects' block sizes to 128KB and splits the large blocks into 128KB
blocks, allowing the large-block filesystem to be received by a system
that does not support the `large_blocks` feature.  A send stream
generated by `zfs send -L` (or `--large-block`) preserves the large
block size on the receiving system, by using large WRITE records.

When receiving an incremental send stream for a filesystem with large
blocks, if the send stream's -L flag was toggled, a bug is encountered
in which the file's contents are incorrectly zeroed out.  The contents
of any blocks that were not modified by this send stream will be lost.
"Toggled" means that the previous send used `-L`, but this incremental
does not use `-L` (-L to no-L); or that the previous send did not use
`-L`, but this incremental does use `-L` (no-L to -L).

Changes:

This commit addresses the problem with several changes to the semantics
of zfs send/receive:

1. "-L to no-L" incrementals are rejected.  If the previous send used
`-L`, but this incremental does not use `-L`, the `zfs receive` will
fail with this error message:

    incremental send stream requires -L (--large-block), to match
    previous receive.

2. "no-L to -L" incrementals are handled correctly, preserving the
smaller (128KB) block size of any already-received files that used large
blocks on the sending system but were split by `zfs send` without the
`-L` flag.

3. A new send stream format flag is added, `SWITCH_TO_LARGE_BLOCKS`.
This feature indicates that we can correctly handle "no-L to -L"
incrementals.  This flag is currently not set on any send streams.  In
the future, we intend for incremental send streams of snapshots that
have large blocks to use `-L` by default, and these streams will also
have the `SWITCH_TO_LARGE_BLOCKS` feature set. This ensures that streams
from the default use of `zfs send` won't encounter the bug mentioned
above, because they can't be received by software with the bug.

Implementation notes:

To facilitate accessing the ZPL's generation number,
`zfs_space_delta_cb()` has been renamed to `zpl_get_file_info()` and
restructured to fill in a struct with ZPL-specific info including owner
and generation.

In the "no-L to -L" case, if this is a compressed send stream (from
`zfs send -cL`), large WRITE records that are being written to small
(128KB) blocksize files need to be decompressed so that they can be
written split up into multiple blocks.  The zio pipeline will recompress
each smaller block individually.

A new test case, `send-L_toggle`, is added, which tests the "no-L to -L"
case and verifies that we get an error for the "-L to no-L" case.

Reviewed-by: Paul Dagnelie <pcd@delphix.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Matthew Ahrens <mahrens@delphix.com>
Closes #6224 
Closes #10383

nivedita76 committed:
Cleanup libzpool/kernel.c

Commit
  ec213971274a ("async zvol minor node creation interferes with receive")
replaced zvol_create_minors with zvol_create_minor and
zvol_create_minors_recursive, changing the prototype at the same time.

However the stub functions in libzpool/kernel.c were defined with the
old prototype. As the definitions are empty, this doesn't cause any
runtime issues, but an LTO build shows warnings because of the
mismatched prototypes.

Commit
  a0bd735adb1b ("Add support for asynchronous zvol minor operations")
removed the real zvol_remove_minor, but for some reason added a stub
implementation in libzpool/kernel.c with no references. Delete this dead
code.

Commit
  196bee4cfd57 ("Remove deduplicated send/receive code")
removed zfs_onexit_del_cb and zfs_onexit_cb_data. Drop the stubs as
well.

Add zvol.h include to provide prototypes, and sort the include
directives.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

Gelma committed:
Fix typos

Correct various typos in the comments and tests.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Andrea Gelmini <andrea.gelmini@gelma.net>
Closes #10423

Commit:
Fixup "Avoid the GEOM topology lock recursion when autoexpanding a pool"

The patch was applied to vdev_geom_open instead of vdev_geom_close by
mistake.

Reviewed-by: Alexander Motin <mav@FreeBSD.org>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #10427

ahrens committed:
Remove unnecessary references to slavery

The horrible effects of human slavery continue to impact society.  The
casual use of the term "slave" in computer software is an unnecessary
reference to a painful human experience.

This commit removes all possible references to the term "slave".

Implementation notes:

The zpool.d/slaves script is renamed to dm-deps, which uses the same
terminology as `dmsetup deps`.

References to the `/sys/class/block/$dev/slaves` directory remain.  This
directory name is determined by the Linux kernel.  Although
`dmsetup deps` provides the same information, it unfortunately requires
elevated privileges, whereas the `/sys/...` directory is world-readable.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Signed-off-by: Matthew Ahrens <mahrens@delphix.com>
Closes #10435

ggzengel committed:
bash_completion: add missing attributes

There a some attributes missing which are shown in man pages:
zfs list -t type
           A comma-separated list of types to display, where type is one of filesystem, snapshot, volume, *bookmark*, or all.  For example, specifying -t snapshot displays only snapshots.
zfs get -s source
           A comma-separated list of sources to display.  Those properties coming from a source other than those in this list are ignored.  Each source must be one of the following: local, default, inherited, temporary, *received*, and none.  The default value is all sources.
zfs get -t type
           A comma-separated list of types to display, where type is one of filesystem, snapshot, volume, bookmark, or all.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Grischa Zengel <github.zfsonlinux@zengel.info>
Closes #10418

ggzengel committed:
man.8: Add bookmark to list of types

While checking bash_completion I missed bookmark as type.

```
# zfs get type zpool2#b
NAME      PROPERTY  VALUE     SOURCE
zpool2#b  type      bookmark  -
```

Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: George Melikov <mail@gmelikov.ru>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Grischa Zengel <github.zfsonlinux@zengel.info>
Closes #10419

bwatkinson committed:
Removing ZERO_PAGE abd_alloc_zero_scatter

For MIPS architectures on Linux the ZERO_PAGE macro references
empty_zero_page, which is exported as a GPL symbol. The call to
ZERO_PAGE in abd_alloc_zero_scatter has been removed and a single
zero'd page is now allocated for each of the pages in abd_zero_scatter
in the kernel ABD code path.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Brian Atkinson <batkinson@lanl.gov>
Closes #10428

nivedita76 committed:
Add include files for prototypes

Include the header with prototypes in the file that provides definitions
as well, to catch any mismatch between prototype and definition.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

nivedita76 committed:
Add prototypes

Add prototypes/move prototypes to header files.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

lundman committed:
macOS: rebase master fixes

lundman committed:
Upstream: Changes to zfs-tests [squash]

lundman committed:
macOS: Add all tunables from upstream

Commit:
FreeBSD: Don't require zeroing new locks before init

This has not shown to be of use enough to justify the inconvenience.

Reviewed-by: Matt Macy <mmacy@FreeBSD.org>
Reviewed-by: Alexander Motin <mav@FreeBSD.org>
Reviewed-by: Allan Jude <allanjude@freebsd.org>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #10449

gamanakis committed:
Fix gcc10.1 truncation error

gcc10.1 complains with:

../../include/sys/dmu.h:373:24: error: ‘%s’ directive output may be
truncated writing up to 95 bytes into a region of size 75
[-Werror=format-truncation=]
  373 | #define DMU_POOL_DDT   "DDT-%s-%s-%s"
      |                        ^~~~~~~~~~~~~~
../../module/zfs/ddt.c:256:37: note: in expansion of macro
‘DMU_POOL_DDT’
  256 |  (void) snprintf(name, DDT_NAMELEN, DMU_POOL_DDT,
      |                                     ^~~~~~~~~~~~
../../include/sys/dmu.h:373:32: note: format string is defined here
  373 | #define DMU_POOL_DDT   "DDT-%s-%s-%s"
      |                                ^~
../../module/zfs/ddt.c:256:9: note: ‘snprintf’ output 7 or more bytes
(assuming 102) into a destination of size 80
  256 |  (void) snprintf(name, DDT_NAMELEN, DMU_POOL_DDT,
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  257 |      zio_checksum_table[ddt->ddt_checksum].ci_name,
      |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  258 |      ddt_ops[type]->ddt_op_name, ddt_class_name[class]);
      |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Increasing DTT_NAMELEN fixes it.

Reviewed-By: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: George Amanakis <gamanakis@gmail.com>
Closes #10433

lundman committed:
Upstream: zil_commit_waiter() can stall forever

On macOS clock_t is unsigned, so when cv_timedwait_hires() returns -1
we loop forever. The conditional was tweaked to ignore signedness.

Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10445

lundman committed:
Add convenience wrappers for common uio usage

The macOS uio struct is opaque and the API must be used, this
makes the smallest changes to the code for all platforms.

Reviewed-by: Matt Macy <mmacy@FreeBSD.org>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10412

lundman committed:
macOS: Allow sys/vdev_disk.h to have OS fields

nivedita76 committed:
Remove dead code

Delete unused functions.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

nivedita76 committed:
Mark functions as static

Mark functions used only in the same translation unit as static. This
only includes functions that do not have a prototype in a header file
either.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

lundman committed:
macOS: compile fixes after vdev_disk_t change

lundman committed:
macOS: remove semaphore.h

lundman committed:
Revert "Upstream: move asm_linkage.h to platform folder"

This reverts commit 9765d8ec2a9749b71d60092b730e0cb0e428a73f.

lundman committed:
Upstream: ASM changes to support macOS

lundman committed:
macOS: ASM changes after revert

Commit:
Fix FreeBSD condvar semantics

We should return -1 instead of negative deltas, and 0 if signaled.

Reviewed-by: Alexander Motin <mav@FreeBSD.org>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Jorgen Lundman <lundman@lundman.net>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #10460

bwatkinson committed:
Fixing ABD struct allocation for FreeBSD

In the event we are allocating a gang ABD in FreeBSD we are passing 0
to abd_alloc_struct(); however, this led to an allocation of ABD scatter
with 0 chunks. This left the gang ABD allocation 24 bytes smaller than
it should have been.

Reviewed-by: Matt Macy <mmacy@FreeBSD.org>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Co-authored-by: Matt Macy <mmacy@FreeBSD.org>
Signed-off-by: Brian Atkinson <batkinson@lanl.gov>
Closes #10431

ahrens committed:
Remove refences to blacklist/whitelist

These terms reinforce the incorrect notion that black is bad and white
is good.

Replace this language with more specific terms which are also more clear
and don't rely on metaphor.  Specifically:

* When vdevs are specified on the command line, they are the "selected"
vdevs.

* Entries in /dev/ which should not be considered as possible disks are
"excluded" devices.

Reviewed-by: John Kennedy <john.kennedy@delphix.com>
Reviewed-by: Ryan Moeller <ryan@ixsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: George Wilson <gwilson@delphix.com>
Signed-off-by: Matthew Ahrens <mahrens@delphix.com>
Closes #10457

lundman committed:
Make struct vdev_disk_t be platform private

Linux defines different vdev_disk_t members to macOS, but they are
only used in vdev_disk.c so move the declaration there.

Reviewed-by: Ryan Moeller <ryan@ixsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10452

Commit:
FreeBSD: Kernel module should depend on xdr not krpc after 1300092

Since https://reviews.freebsd.org/D24408 FreeBSD provides XDR functions
in the xdr module instead of krpc.

For FreeBSD 13, the MODULE_DEPEND should be changed to xdr

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #10442 
Closes #10443

lundman committed:
drr_begin: can't forward declare untagged struct

When compiling with Clang++ it does not allow for untagged structs, so
struct ddr_begin needs to be declared before the struct that uses it.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10453

dioni21 committed:
Merge bash_completions changes from upstream

The current bash_completion contrib code in openzfs is very old, and
some changes have been added since.

The original repo is at https://github.com/Aneurin/zfs-bash

I've been using the original @Aneurin code since my first deploy of ZoL.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: João Carlos Mendes Luís <jonny@jonny.eng.br>
Closes #10456

nivedita76 committed:
Make Skein_{Get,Put}64_LSB_First inline functions

Turn the generic versions into inline functions and avoid
SKEIN_PORT_CODE trickery.

Also drop the PLATFORM_MUST_ALIGN check for using the fast bcopy
variants. bcopy doesn't assume alignment, and the userspace version is
currently different because the _ALIGNMENT_REQUIRED macro is only
defined by the kernelspace headers.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

nivedita76 committed:
Enable -Wmissing-prototypes/-Wstrict-prototypes

Switch on warning flags to detect mismatch between declaration and
definition.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

lundman committed:
macOS: correct cv_wait_sig() return values

lundman committed:
macOS: Ensure to use -O2 on dsl_scan.c

lundman committed:
Upstream: zfs-tests script

nivedita76 committed:
Switch off -Wmissing-prototypes for libgcc math functions

spl-generic.c defines some of the libgcc integer library functions on
32-bit. Don't bother checking -Wmissing-prototypes since nothing should
directly call these functions from C code.

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Arvind Sankar <nivedita@alum.mit.edu>
Closes #10470

amotin committed:
Set initial arc_c to arc_c_min instead of arc_c_max

For at least 15 years since OpenSolaris arc_c was set by default to
arc_c_max, later decreased under memory pressure.  I've noticed that
if arc_c was set high enough to cause memory pressure as considered
by ZFS, setting of arc_no_grow to TRUE in arc_reap_cb_check() makes
no effect until both arc_kmem_reap_soon() and delay(reap_retry_ms)
return.  All that time ZFS can continue increasing its effective ARC
size, causing more memory pressure, potentially up to the point when
OS low memory handler activates and reduces arc_c, requesting fast
reclamation of just allocated memory.

The problem seems to be more serious on FreeBSD and I guess Linux,
since neither of them implement/use asynchronous kmem reclamation,
so arc_kmem_reap_soon() can take more time.  On older FreeBSD 11 not
supporting multiple memory domains system with lots of RAM can get
completely unresponsive for minutes due to heavy lock congestion
between ARC reclamation and page daemon kmem reclamation threads.
With this change to more conservative arc_c value ARC stops growing
just it time and does not need later reclamation.

Also while there, since now growing arc_c is a more often situation,
use aggsum_upper_bound() instead of aggsum_compare() in arc_adapt()
to reduce lock congestion.  It is also getting in sync with code in
arc_get_data_impl().

Reviewed-by: Ryan Moeller <ryan@iXsystems.com>
Reviewed-by: Allan Jude <allanjude@freebsd.org>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Alexander Motin <mav@FreeBSD.org>
Sponsored-By: iXsystems, Inc.
Closes #10437

lundman committed:
zfs_ioctl: saved_poolname can be truncated

As it uses kmem_strdup() and kmem_strfree() which both rely on
strlen() being the same, but saved_poolname can be truncated causing:

SPL: kernel memory allocator:
buffer freed to wrong cache
SPL: buffer was allocated from kmem_alloc_16,
SPL: caller attempting free to kmem_alloc_8.
SPL: buffer=0xffffff90acc66a38  bufctl=0x0  cache: kmem_alloc_8

Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10469

mattmacy committed:
Add abd_cache_reap_now for abd_chunk_cache users

Apparently missed in the initial port integration was
the need to reap the abd_chunk_cache on FreeBSD. This
change addresses that oversight.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Alexander Motin <mav@FreeBSD.org>
Signed-off-by: Matt Macy <mmacy@FreeBSD.org>
Closes #10474

lundman committed:
macOS: zed changes

lundman committed:
Upstream: addition zfstests work

lundman committed:
even more zfs-tests

lundman committed:
zfs-tests

mattmacy committed:
Disambiguate condvar API contract

On Illumos callers of cv_timedwait and cv_timedwait_hires
can't distinguish between whether or not the cv was signaled
or the call timed out. Illumos handles this (for some definition
of handles) by calling cv_signal in the return path if we were
signaled but the return value indicates instead that we timed
out. This would make sense if it were possible to query the the
cv for its net signal disposition. However, this isn't possible
and, in spite of the fact that there are places in the code that
clearly take a different and incompatible path if a timeout value
is indicated, this distinction appears to be rather subtle to most
developers. This problem is further compounded by the fact that on
Linux, calling cv_signal in the return path wouldn't even do the
right thing unless there are other waiters.

Since it is possible for the caller to independently determine how
much time is remaining but it is not possible to query if the cv
was in fact signaled, prioritizing signalling over timeout seems
like a cleaner solution. In addition, judging from usage patterns
within the code itself, it is also less error prone.

Reviewed-by: Jorgen Lundman <lundman@lundman.net>
Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Alexander Motin <mav@FreeBSD.org>
Signed-off-by: Matt Macy <mmacy@FreeBSD.org>
Closes #10471

lundman committed:
Avoid adding new primitives in zpool wait

zpool wait brought in sem_init() and family, which is a primitive set
not previously used in Open ZFS. It also happens to be deprecated
on macOS. Replace with phtread API calls.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: John Gallagher <john.gallagher@delphix.com>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10468

adilger committed:
linux: add basic fallocate(mode=0/2) compatibility

Implement semi-compatible functionality for mode=0 (preallocation)
and mode=FALLOC_FL_KEEP_SIZE (preallocation beyond EOF) for ZPL.

Since ZFS does COW and snapshots, preallocating blocks for a file
cannot guarantee that writes to the file will not run out of space.
Even if the first overwrite was guaranteed, it would not handle any
later overwrite of blocks due to COW, so strict compliance is futile.
Instead, make a best-effort check that at least enough free space is
currently available in the pool (with a bit of margin), then create
a sparse file of the requested size and continue on with life.

This does not handle all cases (e.g. several fallocate() calls before
writing into the files when the filesystem is nearly full), which
would require a more complex mechanism to be implemented, probably
based on a modified version of dmu_prealloc(), but is usable as-is.

A new module option zfs_fallocate_reserve_percent is used to control
the reserve margin for any single fallocate call.  By default, this
is 110% of the requested preallocation size, so an additional 10% of
available space is reserved for overhead to allow the application a
good chance of finishing the write when the fallocate() succeeds.
If the heuristics of this basic fallocate implementation are not
desirable, the old non-functional behavior of returning EOPNOTSUPP
for calls can be restored by setting zfs_fallocate_reserve_percent=0.

The parameter of zfs_statvfs() is changed to take an inode instead
of a dentry, since no dentry is available in zfs_fallocate_common().

A few tests from @behlendorf cover basic fallocate functionality.

Reviewed-by: Richard Laager <rlaager@wiktel.com>
Reviewed-by: Arshad Hussain <arshad.super@gmail.com>
Reviewed-by: Matthew Ahrens <mahrens@delphix.com>
Co-authored-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Andreas Dilger <adilger@dilger.ca>
Issue #326
Closes #10408

Commit:
Match new vfs_checkexp KPI in FreeBSD head

KPI changed in FreeBSD, update accordingly.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #10475

lundman committed:
macOS: Disambiguate condvar API contract

This is the macOS changes required for the upstream commit:
8056a75672a57c85b8e10c0c6bce138146f7d213

Also correct cstyle.

remember cv_timedwait() and cv_timedwait_hires() take different
timeout types!

lundman committed:
macOS: Add abd_cache_reap_now for abd_chunk_cache users

macOS changes needed for commit 7564073ed6344c12e6bc4ffabd130522d937fb93

lundman committed:
Upstream: zfstests run large.run

lundman committed:
Also include said large.run file

lundman committed:
Take out test which panic

lundman committed:
macOS: fix zfstester get_time

lundman committed:
zfs-tests: update large.run, make smaller macOS-CI.run

In the hopes that it finishes in 60 mins.

lundman committed:
make zfs-tests shorter for CI

lundman committed:
even smaller! we have the technology!

lundman committed:
update macos=CI.run

lundman committed:
macOS: can not pass componentname to ZFS

as ZFS updates the name with the real name (case insensitive) and
will cause panic, as it is often protected. We need to pass a buffer
to use, and use the returned name as cachename.

lundman committed:
macOS: remove more tests to fit in 60mins

lundman committed:
zed additional features

This commit adds two features to zed, that macOS desires. The first
is that when you unload the kernel module, zed would enter into a
cpubusy loop calling zfs_events_next() repeatedly. We now look for
ENODEV, returned by kernel, so zed can exit gracefully.

Second feature is -I (idle) (alas -P persist was taken) is for the
deamon to;

1; if started without ZFS kernel module, stick around waiting for it.
2; if kernel module is unloaded, go back to 1.

This is due to daemons in macOS is started by launchctl, and is
expected to stick around.

Currently, the busy loop only exists when errno is ENODEV. This is
to ensure that functionality that upstream expects is not changed.
It did not care about errors before, and it still does not. (with the
exception of ENODEV).

However, it is probably better that all errors
(ERESTART notwithstanding) exits the loop, and the issues complaining
about zed taking all CPU will go away.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Jorgen Lundman <lundman@lundman.net>
Closes #10476

lundman committed:
macOS: zvol need to verify and lock zv before use.

When called from, or calling to C++, we need to treat the stored 'zv'
as a unique ID, and verify it is still in the list of active zvols.
If found, lock in the correct order.

lundman committed:
SPL: correct cstyle.pl

arun-kv committed:
Fixed kmem_bufctl_cache memory leak (#273)

lundman committed:
macOS: cstyle fixes, remove spl-kobj.c

lundman committed:
macOS: cstyle.pl-ify SPL

lundman committed:
macOS: cstyle.pl all kernel header files.

lundman committed:
macOS: cstyle kernel C files

lundman committed:
macOS: cstyle kernel CPP files

lundman committed:
macOS: libspl cstyle fixes.

arun-kv committed:
Fixed memory leak during zfs driver unload. (#279)

lundman committed:
macOS: fix file based pools

zfs_dev_is_whole_disk() was incorrectly ported from FreeBSD. Use
Linux version.

lundman committed:
macOS: undo previous decmpfs handling

We used to handle decmpfs by uncompressing it after the fact, but it
added a lot of dependencies.

Now, if decmpfs XATTR is set, AND it sets UF_COMPRESSED, AND is
followed by a setsize(0), we will ignore the truncate, and delete
the XATTR.

Only concern is if we were to stop a truncate, which was supposed to happen
(ie, not decmpfs related).

lundman committed:
macOS: make codecheck pass

lundman committed:
Upstream: skip cstyle on kextsymboltool.c

lundman committed:
macOS: change send/recv pipe to fifo

Simplify kernel dependencies by only do IO on vnodes, and userland
to wrap send to pipes with fifo (opened to get vnode) and relay
child processes.

lundman commented:
This PR is no longer needed.
